### ISSUE TITLE
Reference implementation of conditional observe parameters 

### ISSUE THREAD

**Issue by mjkoster** [core-wg/dynlink#10]
Hi,

I made some code last weekend that includes a new implementation of the observe parameters and which includes the new "band" parameter.

It processes scalar (number) values, strings, and booleans by type. 

It is only a little tested so far, and I haven't created any formal test cases. 

There is a bigger work in progress that shows the context of this:
https://github.com/sensornetworksystems/sns-monitor-embedded/tree/master/arduino/sensor

Please let me know what you think. I think we can make a plantUML state diagram for this. I'm not sure how to normatively specify the rules. 

```
unsigned int apply_conditionals (Resource * resource) {
  /*
   * Returns true if a notification on the resource is to be scheduled
   * last_sample_time is the timestamp of the sample being processed
   * last_rep_time is the timestamp of the sample last reported in a notification
   * v is scalar value of the sample being processed
   * vs is string value
   * vb is boolean value
   * last_rep_v (vb, vs) hold the most recent previously reported value 
   */

  // if pmin is not yet exceeded, return indicating no notification
  if (resource->last_sample_time - resource->last_rep_time < resource->pmin) {
    return(false);
  }
  if (num_type == resource->type) {
    // notify on gt crossings if not in band mode, gt == vmax to disable gt limiting
    if (  (false == resource->band) && (resource->gt < resource->vmax) &&
    ( (resource->v > resource->gt && resource->last_rep_v <= resource->gt) ||
    (resource->v <= resource->gt && resource->last_rep_v > resource->gt) ) )
    {
      resource->last_rep_v = resource->v;
      return(true);
    }
    // notify on lt crossings if not in band mode, lt == vmin to disable lt limiting
    else if ( (false == resource->band) && (resource->lt > resource->vmin) &&
    ( (resource->v < resource->lt && resource->last_rep_v >= resource->lt) ||
    (resource->v >= resource->lt && resource->last_rep_v < resource->lt) ) )
    {
      resource->last_rep_v = resource->v;
      return(true);
    }
    // notify on either step or pmax exceeded if not in band mode
    // set step == 0 to disable value filtering
    else if ( false == resource->band &&
      ( ( resource->v - resource->last_rep_v >= resource->st
      || resource->last_rep_v - resource->v >= resource->st )
      || (resource->last_sample_time - resource->last_rep_time >= resource->pmax) ) )
    {
      resource->last_rep_v = resource->v;
      return(true);
    }
    // notify on either step or pmax exceeded if in band mode and in band
    // set step == 0 to disable value filtering
    else if ( true == resource->band && (
      ( (resource->v >= resource->gt) && (resource->v <= resource->lt) )
      ||( (resource->v >= resource->gt) && (resource->gt >= resource->lt) )
      || ( (resource->v <= resource->lt) && (resource->gt >= resource->lt) ) ) // in band
      && ( ( resource->v - resource->last_rep_v >= resource->st
        || resource->last_rep_v - resource->v >= resource->st )
        || (resource->last_sample_time - resource->last_rep_time >= resource->pmax) ) )
    {
      resource->last_rep_v = resource->v;
      return(true);
    }
    else
      // no numeric value condition to notify
      return(false);
  }
  else {
    // evaluate value change and pmax for boolean and string types
    if ( resource->vb != resource->last_rep_vb ) {
      resource->last_rep_vb = resource->vb;
      return(true);
    }
    else if ( resource->vs != resource->last_rep_vs ) {
      resource->last_rep_vs = resource->vs;
      return(true);
    }
    else if (resource->last_sample_time - resource->last_rep_time >= resource->pmax) {
      resource->last_rep_time = resource->last_sample_time;
      return(true); // no value change
    }
    else
      // no notifiable value type
      return(false);
  }
}
```

**Comment by mjkoster** [comment]
Here's a refactored logic version that's easier to read. The variables and assignments could be replaced by #defines to save some CPU time and memory.

```
//notifiable.c
bool notifiable( Resource * r ) {

  bool scalar_type = ( num_type == r->type ); // number, string. or boolean
  bool string_type = ( str_type == r->type );
  bool boolean_type = ( bool_type == r->type );
  time_t t = r->last_sample_time; // when v was sampled
  time_t t_last = r->last_rep_time; // last time this resource was reported
  double v = r->v; // sampled value
  char* vs = r->vs;
  bool vb = r->vb;
  double v_last = r->last_rep_v; // last reported value
  char* vs_last = r->last_rep_vs;
  bool vb_last = r->last_rep_vb;
  time_t pmin = r->pmin; // min time between report notifications
  time_t pmax = r->pmax; // max time allowed without a notificaiton
  double lt = r->lt; // limit for sample less than
  double gt = r->gt; // limit for sample greater than
  double st = r->st; // limit of sample change (delta) since last report
  bool band = r->band; // only notify when gt <= sample <= lt

  bool pmin_ex = ( t - t_last >= pmin ); // pmin exceeded
  bool pmax_ex = ( t - t_last > pmax ); // pmax exceeded
  bool lt_ex = ( v < lt ^ v_last < lt ); // lt crossed
  bool gt_ex = ( v > gt ^ v_last > gt ); // gt crossed
  bool st_ex = ( abs( v - v_last ) >= st ); // st value exceeded from last notification
  bool in_band = ( ( gt <= v && v <= lt ) || 
                ( v >= gt && gt >= lt ) ||
                ( v <= lt && lt <= gt ) ); // if the sample is in-band

  return (
    pmin_ex &&
    ( scalar_type ?
      ( ( !band && ( gt_ex || lt_ex || st_ex || pmax_ex ) ) ||
        ( band && in_band && ( st_ex || pmax_ex) ) )
    : string_type ?
      ( ( vs != vs_last) || pmax_ex )
    : boolean_type ?
      ( ( vb != vb_last) || pmax_ex )
    : false )
  );
}
```

**Comment by mjkoster** [comment]
And here's what I ended up with:

```
//notifiable.c
bool notifiable() {

  enum { NUMBER_TYPE, STRING_TYPE, BOOLEAN_TYPE } type;
  time_t t; // when v was sampled
  time_t t_last; // last time this resource was reported
  double v; // sampled number value
  char* vs; // sampled string value
  bool vb; // sampled boolean value
  double v_last; // last reported number value
  char* vs_last; // last reported string value
  bool vb_last; // last reported boolean value
  time_t pmin; // min time between report notifications
  time_t pmax; // max time allowed without a notificaiton
  double lt; // limit for sample less than
  double gt; // limit for sample greater than
  double st; // limit of sample change (delta) since last report
  bool band; // only notify when gt <= sample <= lt

  #define BAND band
  #define SCALAR_TYPE ( NUMBER_TYPE == type )
  #define STRING_TYPE ( STRING_TYPE == type )
  #define BOOLEAN_TYPE ( BOOLEAN_TYPE == type )
  #define PMIN_EX ( t - t_last >= pmin )
  #define PMAX_EX ( t - t_last > pmax )
  #define LT_EX ( v < lt ^ v_last < lt )
  #define GT_EX ( v > gt ^ v_last > gt )
  #define ST_EX ( abs( v - v_last ) >= st )
  #define IN_BAND ( ( gt <= v && v <= lt ) || \
                  ( lt <= gt && gt <= v ) || \
                  ( v <= lt && lt <= gt ) )
  #define VB_CHANGE ( vb != vb_last )
  #define VS_CHANGE ( vs != vs_last )

  return (
    PMIN_EX &&
    ( SCALAR_TYPE ?
      ( ( !BAND && ( GT_EX || LT_EX || ST_EX || PMAX_EX ) ) ||
        ( BAND && IN_BAND && ( ST_EX || PMAX_EX) ) )
    : STRING_TYPE ?
      ( VS_CHANGE || PMAX_EX )
    : BOOLEAN_TYPE ?
      ( VB_CHANGE || PMAX_EX )
    : false )
  );
}
```

**Comment by mjkoster** [comment]
Using the Resource structure defined in sensor.c 
```
bool notifiable( Resource * r ) {

#define BAND r->band
#define SCALAR_TYPE ( num_type == r->type )
#define STRING_TYPE ( str_type == r->type )
#define BOOLEAN_TYPE ( bool_type == r->type )
#define PMIN_EX ( r->last_sample_time - r->last_rep_time >= r->pmin )
#define PMAX_EX ( r->last_sample_time - r->last_rep_time > r->pmax )
#define LT_EX ( r->v < r->lt ^ r->last_rep_v < r->lt )
#define GT_EX ( r->v > r->gt ^ r->last_rep_v > r->gt )
#define ST_EX ( abs( r->v - r->last_rep_v ) >= r->st )
#define IN_BAND ( ( r->gt <= r->v && r->v <= r->lt ) || ( r->lt <= r->gt && r->gt <= r->v ) || ( r->v <= r->lt && r->lt <= r->gt ) )
#define VB_CHANGE ( r->vb != r->last_rep_vb )
#define VS_CHANGE ( r->vs != r->last_rep_vs )

  return (
    PMIN_EX &&
    ( SCALAR_TYPE ?
      ( ( !BAND && ( GT_EX || LT_EX || ST_EX || PMAX_EX ) ) ||
        ( BAND && IN_BAND && ( ST_EX || PMAX_EX) ) )
    : STRING_TYPE ?
      ( VS_CHANGE || PMAX_EX )
    : BOOLEAN_TYPE ?
      ( VB_CHANGE || PMAX_EX )
    : false )
  );
}
```

**Comment by bsilverajan** [comment]
I think a UML diagram can be very useful. We could place it into section 4.7 on Attribute Interactions. Could we include the code as an Appendix to the document?

**Comment by mjkoster** [comment]
Yes, I made a version of the reference code that uses mostly boolean logic and descriptive variables that would be good fto include as documentation. I think it implements the correct behavior we discussed.

**Comment by bsilverajan** [comment]
I've included the code into Section 4.7. When the state machine diagram is ready, it can also be included there.

**Comment by bsilverajan** [comment]
This issue has largely been superceded by editorial changes and the addition of extra examples to clarify the conditional observe parameters. A state machine diagram as such won't provide significant benefit, since the examples, combined with the reference code in the draft, should be self-explanatory
